modules/extensions-support/src/main/java/org/mule/module/extension/internal/model/property/SubtypesModelProperty.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.module.extension.internal.model.property;
+
+
+import org.mule.extension.api.introspection.ModelProperty;
+import org.mule.module.extension.internal.util.IntrospectionUtils;
+import org.mule.util.CollectionUtils;
+
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+
+/**
+ * An immutable model property which specifies the declared implementations
+ * for the {@link ModelProperty}'s holder type
+ *
+ * @since 4.0
+ */
+public final class SubtypesModelProperty implements ModelProperty
+{
+
+    private final List<Class<?>> subtypes;
+
+    /**
+     * Creates a new instance containing all the declared {@link #subtypes} for
+     * the {@link ModelProperty}'s owner base type
+     *
+     * @param subTypes subtypes of the {@link ModelProperty}'s holder base type
+     */
+    public SubtypesModelProperty(List<Class<?>> subTypes)
+    {
+
+        if (CollectionUtils.isEmpty(subTypes))
+        {
+            throw new IllegalArgumentException("Subtypes declaration must have at least one referenced type");
+        }
+
+        for (Class<?> subtype : subTypes)
+        {
+            if (!IntrospectionUtils.isInstantiable(subtype))
+            {
+                throw new IllegalArgumentException(String.format(
+                        "All declared Subtypes have to be instantiable, but %s isn't", subtype.getName()));
+            }
+        }
+
+        this.subtypes = subTypes;
+    }
+
+    /**
+     * @return the sub types declared for the {@link ModelProperty}'s holder type
+     */
+    public List<Class<?>> getSubtypes()
+    {
+        return ImmutableList.copyOf(subtypes);
```

Comment:
don't create an immutable copy every time. Just keep an immutable copy on the constructor, otherwise this is not immutable